### ISSUE TITLE
Wikka REST API

### ISSUE THREAD

**Issue by oemunoz** [wikkawik/WikkaWiki#1247]
This is a PHP micro framework that implements Slim, helps you quickly write simple yet powerful web applications and APIs for WikkaWiki.

[wikka-api](https://github.com/pepitosoft/wikka-api) First test Release!!!

- [ ] TODO: Add information from the original README.md
- [ ] TODO: Chosse a correct path on the WikkaWiki estructure.
- [ ] TODO: Implement this over the https port.
- [ ] TODO: Read the configuration from the wikka.config.php.
- [ ] TODO: Versions for the API access. (v1, v2, etc.)
- [x] TODO: Create a temporal token (two hours) usning the user and password from the wikka database.
- [x] TODO: The content of each page must be exported using base64, this make posible avoid stranger charsets.
- [x] TODO: Loggin activity.
- [x] TODO: Support Mysql/sqlite.

**Comment by klenwell** [comment]
I really like the idea of an API. But this is moving awfully fast. Especially when Wikka still doesn't have an automated test suite.

Still, I don't want to be party-pooper. So a couple possible suggestions:

- Somehow modularize all this so it can be installed independently to complement the wikka-core. (That is, turn it into some kind of plugin or extension).
- Or maybe just fork it, as you've effectively done with your new repo, damn the torpedoes, and run with it. 

With the second option, I don't think it would ever make back into the master branch here. But it could be a worthwhile experiment and who knows, maybe I'll be proved wrong.

**Comment by oemunoz** [comment]
@klenwell you are right, this RESTful API is designed to assist on the development process, and it is independent of the rest of the WikkaWiki code too. 

- In essence it only needs to clone the API repo, and edit the src/settings.php file on the API code . **You dont have to do any change on the WikkaWiki code**.
- For now, the best way to get the API is clone the repo some place onto the WikkaWiki directory, and setup the database connection. But **you can also create a new VirtualHost** (preferred https port) and work from there.

This is a very new Feature, needs a lot of work, but works well and has everything that is expected of an API, I do not recommend using it in production or integrate it in version 1.4.0 for now, but surely for a later version it will be a mature feature.

But if you are a developer in a development environment I would appreciate your opinion.

**Comment by chriscase** [comment]
I"m still familiarizing myself with the code, so take what I say with a
grain of salt; but I do think having more of a separation of concerns would
be very helpful.  I have been a bit spoiled at my day job, having
everything really nicely separated (and not in PHP!).  We've got a nice
NodeJS backend with Hapi for our REST API and the front-end is in Angular
JS.  If we could move this project in that direction, it'd be easier to
work on.

**Comment by oemunoz** [comment]
@chriscase Hapi seems to be a very good option to create an API from the beginning, and have to said that I love to work with nodejs, but is true that I want to keep the API as a module, but also I expected to use code that already exists. 

We must take advantage of the knowledge acquired by WikkaWiki during previous versions, and also use the WEB interface like the core of the of the implementation. The use of PHP and DB allows to us to use some code from the already tested and good defined on WikkaWiki. Starting a new implementation in a new language can have side effects, and it can also disseminate existing resources. 

But what I know?, maybe on couple of mounts, I start a new API on Hapi (if you dont start one).... becose that rocks :-), but on that moment must be easier becose we gonna to keep the knowledge of this API.

**Comment by chriscase** [comment]
For sure.  I'm not sure it would be a good idea to use Node in this but
that'd be my choice if I were to do an overhaul.

There's got to be a nice way, though, to do a REST API for this app with
PHP.  I'd love to see how that's done, as I've not done that in PHP before
and I think it would help this code.

**Comment by bakoontz** [comment]
It's not clear to me what advantage there would be to having a REST API layer implemented.  This is definitely something we should talk about, so I've blue-skyed it for now.

**Comment by oemunoz** [comment]
Is true, Is for advanced use, for example I have scripts with HTTPie, I GET and POST direct from the terminal. (this is useful for scripts and cron jobs). Also I have multiple instances,  this instances share some pages (to avoid duplicity), I have a plugin with Guzzle for that.... 

It's not clear for me also, but there is the implementation, needs a lot of work.

**Comment by chriscase** [comment]
In my experience, if the server side code is behind a REST layer, it helps
with code organization, separation of concerns, etc. I've gotten used to
it, in my work with NodeJS and AngularJS.

In that environment you write service modules on the web client to
communicate between the browser layer and the server.

In this particular situation, it's probably best to just use the REST API
for external access.  Someone could write a mobile app, for instance.

In the long run, if it's a good comprehensive API, it'd be possible to use
it on a frontend overhaul.  You could, for instance, use a framework on the
front end like AngularJS and leverage all the power of this to make adding
features and maintaining code easier.
